tools/autograd/templates/Functions.cpp

```diff
@@ -717,6 +717,13 @@ Tensor diag_backward(const Tensor & grad, IntList input_sizes, int64_t diagonal)
   return grad_input;
 }
 
+Tensor diagonal_backward(const Tensor & grad, IntList input_sizes, int64_t offset, int64_t dim1, int64_t dim2) {
+  auto grad_input = at::zeros(grad.type(), input_sizes);
+  auto diag = at::diagonal(grad_input, offset, dim1, dim2);
+  diag.copy_(grad);
```

Comment:
Question: does this work with double backprop, as this is an in place ok?

Reply:
`gradgradcheck(lambda x: torch.diagonal(x,1,2), inputs=(torch.randn(2,3,4),))` returns True. Does that count as working?

Reply:
Should be fine then. I asked because I got bitten in the past with I place operations bring backwards, but I think that by default grad_input doesn't require grad so the double backprop should work fine

Reply:
No, it doesn't work.
```
a = torch.randn(5, 6, requires_grad=True)
b = torch.diagonal(a)**2
c = b.sum()
d, = torch.autograd.grad(c,a, retain_graph=True, create_graph=True)
```
now `d.requires_grad` will be zero.
The strange part is that if I exactly copypaste my diagonal code and replace `self` with `grad_in`, it suddenly works (I got the idea that it would work from the non-square case in `diag_backward`).
I have no idea what the difference is between `auto diag = grad_in.diagonal(...)` and doing the calculations locally and `auto diag = grad_in.as_strided(...)`.
How should I deal with that? Move the calculations in a utility function in a header?

Reply:
I'll duplicate the code for now and add a test that d.requires_grad = True in the above.